modello-maven-plugin/src/main/java/org/codehaus/modello/maven/ModelloVelocityMojo.java

```diff
@@ -0,0 +1,116 @@
+package org.codehaus.modello.maven;
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import java.io.File;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Properties;
+import java.util.stream.Collectors;
+
+import org.apache.maven.plugins.annotations.LifecyclePhase;
+import org.apache.maven.plugins.annotations.Mojo;
+import org.apache.maven.plugins.annotations.Parameter;
+import org.codehaus.modello.plugin.velocity.VelocityGenerator;
+
+/**
+ * Creates files from the model using Velocity templates.
+ * <p>
+ * This mojo can be given a list of templates and a list of parameters.
+ * Each template from the {@link #templates} property will be run with the following context:
+ * <ul>
+ *    <li>{@code version}: the version of the model to generate</li>
+ *    <li>{@code model}: the modello model</li>
+ *    <li>{@code Helper}: a {@link org.codehaus.modello.plugin.velocity.Helper} object instance</li>
+ *    <li>any additional parameters specified using the {@link #params} property</li>
+ * </ul>
+ * The output file is controlled from within the template using the {@code #MODELLO-VELOCITY#REDIRECT}
+ * directive. This allows a single template to generate multiple files. For example, the following
+ * directive will redirect further output from the template to a file named
+ * {@code org/apache/maven/api/model/Plugin.java} if the variable {@code package} is set to
+ * {@code org.apache.maven.api.model} and the variable {@code className} is set to {@code Plugin}.
+ * <p>
+ *     {@code #MODELLO-VELOCITY#REDIRECT ${package.replace('.','/')}/${className}.java}
+ */
```

Comment:
@gnodet nice documentation, I'm starting to understand
adding an integration test in https://github.com/codehaus-plexus/modello/tree/master/modello-maven-plugin/src/it will help showing classical use cases

Reply:
@gnodet looking at the only use case I know of https://github.com/apache/maven/tree/master/maven-model/src/main/mdo
I'm worried on the use case: do you really want to copy paste all the generation code from Modello and tweak it by hand?
I understand that it's easy to prototype some updates to the generated java classes, but I fear it will be unmaintainable

Reply:
Those are mostly new generators not supported by modello.  The v4 model is immutable which is not supported, the v3 wraps the immutable v4 model, which is not supported, the reader returns the v4 immutable model , which is not supported, the transformer did not exist in maven 3.x, and the merger was hand written...  So that's not tweaking the modello code, those are really new generators.
So given those are new generators, we'd have to reimplement them all in modello.

Reply:
However, there's still a problem with the current state in that the templates are duplicated instead of being reused.  This need to be fixed.

Reply:
I had started working on an IT a few weeks ago, but my laptop died with the unfinished work on it (and I can't even turn that laptop on anymore....). I'll see if I can give in another try in the coming weeks.